### ISSUE TITLE
How to set gradient accumulation ?

### ISSUE THREAD

**Issue by michael-conrad** [Tomiinek/Multilingual_Text_to_Speech#17]
I have to use small batches, (size 10), because of memory constraints on my GPU.

How do I set the batch count for gradient accumulation to be used over as part of the hyper params?

Or is this automatic calculated based on "ideal batch" / "actual batch" somewhere?

**Comment by Tomiinek** [comment]
The gradient accumulation is not implemented. However, it supports distributed training on multiple GPUs (see the `parallelization` parameter, batches are then divided between the GPUs and processed in parallel).

  To add gradient accumulation, modify the `train` method in `train.py`. Something like this pseudo-code should be enough:

``` python
...

optimizer.zero_grad() 

for i, batch in enumerate(data):    
   # parse batch ...
   # get teacher forcing ratio
   # run the model ...
   # evaluate loss function ...
   # evaluate adversarial classifier accuracy, if present ...
   loss = loss / args.accumulation_size 
   loss.backward()      
   
   if (i+1) % args.accumulation_size == 0:    
        gradient = torch.nn.utils.clip_grad_norm_(model.parameters(), hp.gradient_clipping)
        optimizer.step()  
        optimizer.zero_grad()           
        # log training progress ...
        # update criterion states (params and decay of the loss and so on ...) ...
       ...
...
```

**Comment by michael-conrad** [comment]
Making sure I'm reading this correctly:

I should add a new arg in __main__:

```python
parser.add_argument("--accumulation_size", type=int, default=1, help="How many batches to perform gradient accumulation over.")
```
and then I should patch the main training loop to read:

```python
def train(logging_start_epoch, epoch, data, model, criterion, optimizer):
    """Main training procedure.
    
    Arguments:
        logging_start_epoch -- number of the first epoch to be logged
        epoch -- current epoch 
        data -- DataLoader which can provide batches for an epoch
        model -- model to be trained
        criterion -- instance of loss function to be optimized
        optimizer -- instance of optimizer which will be used for parameter updates
    """

    model.train() 

    # initialize counters, etc.
    learning_rate = optimizer.param_groups[0]['lr']
    cla = 0
    done, start_time = 0, time.time()

    optimizer.zero_grad() # once before loop start
    # loop through epoch batches
    for i, batch in enumerate(data):     

        global_step = done + epoch * len(data)
        # moved below for gradient accumulation
        #optimizer.zero_grad()  

        # parse batch
        batch = list(map(to_gpu, batch))
        src, src_len, trg_mel, trg_lin, trg_len, stop_trg, spkrs, langs = batch

        # get teacher forcing ratio
        if hp.constant_teacher_forcing: tf = hp.teacher_forcing
        else: tf = cos_decay(max(global_step - hp.teacher_forcing_start_steps, 0), hp.teacher_forcing_steps)

        # run the model
        post_pred, pre_pred, stop_pred, alignment, spkrs_pred, enc_output = model(src, src_len, trg_mel, trg_len, spkrs, langs, tf)
        
        # evaluate loss function
        post_trg = trg_lin if hp.predict_linear else trg_mel
        classifier = model._reversal_classifier if hp.reversal_classifier else None
        loss, batch_losses = criterion(src_len, trg_len, pre_pred, trg_mel, post_pred, post_trg, stop_pred, stop_trg, alignment, 
                                       spkrs, spkrs_pred, enc_output, classifier)

        # evaluate adversarial classifier accuracy, if present
        if hp.reversal_classifier:
            input_mask = lengths_to_mask(src_len)
            trg_spkrs = torch.zeros_like(input_mask, dtype=torch.int64)     
            for s in range(hp.speaker_number):
                speaker_mask = (spkrs == s)
                trg_spkrs[speaker_mask] = s
            matches = (trg_spkrs == torch.argmax(torch.nn.functional.softmax(spkrs_pred, dim=-1), dim=-1))
            matches[~input_mask] = False
            cla = torch.sum(matches).item() / torch.sum(input_mask).item()

        # comptute gradients and make a step
        #loss.backward()      
        #gradient = torch.nn.utils.clip_grad_norm_(model.parameters(), hp.gradient_clipping)
        #optimizer.step()   
        
        # calculate gradient accumulated loss
        loss = loss / args.accumulation_size
        loss.backward()
        
        # apply loss once per accumulation_size batches
        if (i+1) % args.accumulation_size == 0:
            gradient = torch.nn.utils.clip_grad_norm_(model.parameters(), hp.gradient_clipping)
            optimizer.step()
            optimizer.zero_grad()
        
            # log training progress
            if epoch >= logging_start_epoch:
                Logger.training(global_step, batch_losses, gradient, learning_rate, time.time() - start_time, cla) 

            # update criterion states (params and decay of the loss and so on ...)
            criterion.update_states()

            start_time = time.time()
            
        done += 1 
```

I added a zero_grad call before the loop to compensate for it not being called at top of loop anymore.

I added the step to divide the loss by the accumulation_size before doing backward propagation.

The steps from clip_grad_norm_ to start_time set were moved into the modulo if block.
